stdlib/array.mli

```diff
@@ -320,6 +320,71 @@ val of_seq : 'a Seq.t -> 'a array
 (** Create an array from the generator
     @since 4.07 *)
 
+(** {1:array_concurrency Arrays and concurrency safety}
+
+    Care must be taken when concurrently accessing arrays from multiple domains.
+
+    {2:array_atomicity Atomicity}
+
+    Any array operation that accesses more than one element is not guaranteed to
+    be atomic in the presence of concurrent accesses. This includes iteration,
+    scanning, sorting, splitting and combining arrays.
+
+    For example, consider the following program:
+{[let size = 100_000_000
+let a = Array.make size 1
+let d1 = Domain.spawn (fun () -> Array.iteri (fun i x -> a.(i) <- x + 1) a)
+let d2 = Domain.spawn (fun () -> Array.iteri (fun i x -> a.(i) <- 2 * x + 1) a)
+let () = Domain.join d1; Domain.join d2
+]}
+
+    After executing this code, each field of the array [a] is either [2], [3],
+    [4] or [5]. If atomicity is required, then the user must implement their own
+    synchronization (for example, using {!Mutex.t}).
+
+    {2:array_data_race Data races}
+
+    If two domains only access disjoint parts of the array, then the
+    observed behaviour is the equivalent to some sequential interleaving of the
+    operations from the two domains.
+
+    If two domains access the same element of the array, then it is recommended
+    that there is synchronization to mediate the accesses. A data race is said
+    to occur when two domains access the same array element without
+    synchronization and at least one of the accesses is a write.
+
+    In the absence of data races, the observed behaviour is equivalent to some
+    sequential interleaving of the operations from different domains. In the
+    presence of data races, the observed behaviour may not be equivalent to any
+    sequential interleaving of operations from different domains. However, even
+    in the presence of data races, a read operation will return the value of
+    some prior write to that location (with a few exceptions for float arrays).
+    Importantly, programs with data races do not crash.
```

Comment:
I have the impression that there are too much back-and-forth in the three paragraphs. The structure is roughly:

- Disjoint accesses are linearizable
- Non-disjoint accesses should be mediated by synchronization
- Definition of data races
- Data-race free accesses are still linearizable
- With data races, behavior are not linearizable but there are no values out-of-thin-air and no crashes

I would propose to have the description in increasing order of unsafety

- Disjoint accesses are linearizable
- Definition of data races
- Data-race free accesses are still linearizable
- Data races should really be avoided by using synchronization
- With data races, behavior are not linearizable but there are no crashes and no values out-of-thin-air

```
If two domains only access disjoint parts of the array, then the
observed behaviour is the equivalent to some sequential interleaving of the
operations from the two domains.

A data race is said to occur when two domains access the same array element without synchronization
and at least one of the accesses is a write.
In the absence of data races, the observed behaviour is equivalent to some
sequential interleaving of the operations from different domains. 

Whenever possible, data races should be avoided by using synchronization to mediate the accesses to the array elements.
 
Indeed, in the presence of data races, programs will not crash but the observed behaviour may not be equivalent to any
sequential interleaving of operations from different domains. Nevertheless, even
in the presence of data races, a read operation will return the value of some prior write to that location
(with a few exceptions for float arrays).
```